File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2020/Y20D16.kt
package org.gristle.pdxpuzzles.advent.y2020

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.objects.transpose
import org.gristle.pdxpuzzles.utilities.parsing.groupValues

class Y20D16(input: String) : Day {

    data class Rule(val name: String, val lowRange: IntRange, val hiRange: IntRange) {
        fun validFor(values: Iterable<Int>) = values.all { it in lowRange || it in hiRange }
    }

    private val rules = input
        .groupValues("""([a-z]+(?: [a-z]+)?): (\d+)-(\d+) or (\d+)-(\d+)""")
        .map { gv ->
            val lowRange = gv[1].toInt()..gv[2].toInt()
            val hiRange = gv[3].toInt()..gv[4].toInt()
            Rule(gv[0], lowRange, hiRange)
        }

    private val tickets = Regex("""\d+(?:,\d+)+""")
        .findAll(input)
        .toList()
        .map { result -> result.value.split(',').map { it.toInt() } }

    override fun part1(): Int {
        return tickets
            .flatten()
            .filter { value -> rules.all { value !in it.lowRange && value !in it.hiRange } }
            .sum()
    }

    override fun part2(): Long {
        val validTickets = tickets
            .filter { ticket ->
                ticket.all { value -> rules.any { value in it.lowRange || value in it.hiRange } }
            }
        val valueList = validTickets.transpose()
        val sorter = valueList
            .mapIndexed { index, values ->
                index to rules.filter { it.validFor(values) }.toMutableList()
            }.toMutableList()
        val register = mutableMapOf<Rule, Int>()
        while (sorter.isNotEmpty()) {
            sorter.filter { it.second.size == 1 }.forEach { singleton ->
                register[singleton.second.first()] = singleton.first
                sorter.remove(singleton)
                sorter.forEach { it.second.remove(singleton.second.first()) }
            }
        }
        return rules
            .filter { it.name.startsWith("departure") }
            .map { validTickets.first()[register[it]!!] }
            .fold(1L) { acc, i -> acc * i }
    }
}

fun main() = Day.runDay(Y20D16::class)

//    Class creation: 32ms
//    Part 1: 29878 (2ms)
//    Part 2: 855438643439 (8ms)
//    Total time: 42ms

File: rust/advent/src/bin/2020/y20d16.rs
use std::{collections::HashMap, ops::RangeInclusive};

use advent::utilities::get_input::get_input;
use itertools::Itertools;
use utilities::{
    parsing::get_numbers::ContainsNumbers,
    structs::stopwatch::{ReportDuration, Stopwatch},
};
use utilities::structs::grid::Grid2;

type Input<'a> = (Vec<Rule<'a>>, Vec<Vec<usize>>);
type Output = usize;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(20, 16).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

#[derive(Eq, Hash, PartialEq, Clone)]
struct Rule<'a> {
    name: &'a str,
    low: RangeInclusive<usize>,
    high: RangeInclusive<usize>,
}

impl<'a> Rule<'a> {
    fn valid_for(&self, values: &Vec<&usize>) -> bool
    {
        values.into_iter().all(|&it| self.low.contains(it) || self.high.contains(it))
    }
}
fn parse_input(input: &'_ str) -> Input<'_> {
    let (rules, tickets) = input.split_once("\n\n").unwrap();
    let rules = rules
        .lines()
        .map(|rule| {
            let (name, ranges) = rule.split_once(": ").unwrap();
            let (lolo, lohi, hilo, hihi) = ranges.get_numbers().collect_tuple().unwrap();
            Rule {
                name,
                low: lolo..=lohi,
                high: hilo..=hihi,
            }
        })
        .collect();
    let tickets: Vec<Vec<usize>> = tickets
        .lines()
        .map(|ticket| ticket.get_numbers().collect())
        .filter(|ticket: &Vec<usize>| !ticket.is_empty())
        .collect();
    (rules, tickets)
}

fn part1(input: &Input) -> Output {
    let (rules, tickets) = input;
    tickets.iter().flatten()
        .filter(|&value| {
            rules.iter().all(|rule| !rule.low.contains(value) && !rule.high.contains(value))
        })
        .sum() 
}

fn part2(input: &Input) -> Output {
    let (rules, tickets) = input;
    let valid_tickets: Grid2<usize> = tickets.iter()
        .filter(|ticket| ticket.iter().all(|value| {
            rules.iter().any(|rule| rule.low.contains(value) || rule.high.contains(value))
        }))
        .cloned()
        .collect::<Vec<_>>()
        .try_into()
        .unwrap();

    // a list of a tuple. the first value is the field column, the second is all the possible
    // fields the column could correspond to. Gradually eliminate invalid ones, then when there is
    // only one possibility, assign it to the column and take it out of consideration for future
    // field columns.
    let mut sorter: Vec<(usize, Vec<&Rule>)> = valid_tickets
        .columns()
        .enumerate()
        .map(|(index, values)| {
            let filtered: Vec<_> = rules.iter().filter(|rule| rule.valid_for(&values)).collect();
            (index, filtered)
        })
        .collect();
    let mut register: HashMap<&Rule, usize> = HashMap::new();
    while !sorter.is_empty() {
        let (singles, multiples): (Vec<(usize, Vec<&Rule>)>, Vec<(usize, Vec<&Rule>)>) = sorter
            .into_iter()
            .partition(|(_, potential_rules)| potential_rules.len() == 1);
        sorter = multiples;
        for (col_idx, single_vec) in singles {
            for (_, potential_rules) in sorter.iter_mut() {
                potential_rules.retain(|rule| *rule != single_vec[0])
            }
            register.insert(single_vec[0], col_idx);
        }
    }
    rules.iter()
        .filter(|rule| rule.name.starts_with("departure"))
        .map(|rule| valid_tickets.get([register[&rule], 0]).unwrap())
        .fold(1, |acc, i| acc * i)
}

#[test]
fn default() {
    let input = get_input(20, 16).unwrap();
    let input = parse_input(&input);
    assert_eq!(29878, part1(&input));
    assert_eq!(855438643439, part2(&input));
}

// Input parsed (311μs)
// 1. 29878 (20μs)
// 2. 855438643439 (223μs)
// Total: 557μs